### PR TITLE
ROC-1105: Request tracing

### PR DESCRIPTION
**Pull request by rlewan**
### JIRA link

https://tools.hmcts.net/jira/browse/ROC-1105

### Change description

Uses the request tracing capabilities which are being added in https://github.com/hmcts/nodejs-logging/pull/22

### Work checklist

- [x] Unit tests added where applicable

### Developer self-QA run statement

- [x] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: src/main/app/logging/apiLogger.ts [hunk 1]
```diff
@@ -25,7 +31,10 @@ export class ApiLogger {
       message: `API: Response ${responseData.responseCode} from ${responseData.uri} ` +
       ((responseData.responseBody && this.isDebugLevel()) ? `| Body: ${this._stringifyObject(responseData.responseBody)} ` : '') +
       ((responseData.error) ? `| Error: ${this._stringifyObject(responseData.error)} ` : ''),
-      responseCode: responseData.responseCode
+      responseCode: responseData.responseCode,
+      requestId: (responseData.requestHeaders) ? responseData.requestHeaders['Request-Id'] : '',
```
Comment (timja-kainos) [inline src/main/app/logging/apiLogger.ts]: shouldn't these always be set by the middleware?

File: src/main/app/logging/requestTracingHandler.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  private setTracingHeadersInternal (headers: any): void {
+    headers[Headers.ROOT_REQUEST_ID_HEADER] = this.requestTracing.getRootRequestId()
+    headers[Headers.REQUEST_ID_HEADER] = this.requestTracing.createNextRequestId()
+    headers[Headers.ORIGIN_REQUEST_ID_HEADER] = this.requestTracing.getCurrentRequestId()
+  }
+}
+
+function contains (array, value) {
```
Comment (timja-kainos) [inline src/main/app/logging/requestTracingHandler.ts]: array has an `includes` function

File: src/main/app.ts [hunk 3]
```diff
@@ -50,9 +52,9 @@ app.use(bodyParser.urlencoded({
 }))
 app.use(cookieParser())
 
-if (!developmentMode) {
-  app.use(logging.express.accessLogger())
-}
+// if (!developmentMode) {
+app.use(Express.accessLogger())
+// }
```
Comment (rlewan) [inline src/main/app.ts]: I'll revert this before merging.

File: src/main/app/logging/apiLogger.ts (hunk 1 above)
Comment (rlewan) [reply src/main/app/logging/apiLogger.ts]: They should, but I was in a copy-pasting spree at that moment.

File: package.json [hunk 4]
```diff
@@ -30,7 +30,7 @@
     "@hmcts/draft-store-client": "^1.1.1",
     "@hmcts/info-provider": "^1.0.0",
     "@hmcts/nodejs-healthcheck": "^1.4.5",
-    "@hmcts/nodejs-logging": "^1.4.2",
+    "@hmcts/nodejs-logging": "^1.4.3",
```
Comment (rlewan) [inline package.json]: I'll update this to `^2.0.0` once it's published.

File: src/main/app/logging/httpProxyCallInterceptor.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+export const HttpMethods = ['get', 'post', 'put', 'patch', 'delete', 'del', 'head']
+
+export type CallHandler = (target: Object, key: string, callArgs: any[]) => void
+
+export class HttpProxyCallInterceptor {
+  static intercept (target: Object, key: string, handler: CallHandler): any {
```
Comment (timja-kainos) [inline src/main/app/logging/httpProxyCallInterceptor.ts]: it returns a function or something else?
can we do better than `any`?

File: src/main/app/logging/requestPromiseLoggingHandler.ts [hunk 6]
```diff
@@ -1,29 +1,30 @@
...
-  constructor (public request, public apiLogger) {
-    this.request = request
-    this.apiLogger = apiLogger
+  constructor (private request, private apiLogger = new ApiLogger()) {
+    if (!this.request) {
+      throw new Error('Initialised request instance is required')
+    }
+  }
+
+  static proxy (request) {
```
Comment (timja-kainos) [inline src/main/app/logging/requestPromiseLoggingHandler.ts]: can we set types here?

File: src/main/app/logging/requestTracingHandler.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    if (!this.request) {
+      throw new Error('Initialised request instance is required')
+    }
+  }
+
+  static proxy (request) {
+    return new Proxy(request, new RequestTracingHandler(request))
+  }
+
+  get (target, key) {
```
Comment (timja-kainos) [inline src/main/app/logging/requestTracingHandler.ts]: types?
same in general in this file

File: src/main/app/logging/httpProxyCallInterceptor.ts (hunk 5 above)
Comment (rlewan) [reply src/main/app/logging/httpProxyCallInterceptor.ts]: It's a handler for [`Proxy.get`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler/get), could be anything.

**Pull request merged**